libraries/AC_SafeRTL/AC_SafeRTL.h

```diff
@@ -0,0 +1,89 @@
+    Vector3f path[MAX_PATH_LEN];
+    uint8_t _last_index;
+public:
+    bool accepting_new_points; // false means that any call to append_if_far_enough() will fail. This should be unset when entering SafeRTL mode, and set when exiting.
+    SafeRTL_Path(bool);
+    void append_if_far_enough(Vector3f);
+    bool routine_cleanup();
+    Vector3f* thorough_cleanup();
+    Vector3f get(int index) { return path[index]; }
+    bool pop_point(Vector3f&);
+    void reset_path(Vector3f);
+    bool cleanup_ready() { return _pruning_complete && _simplification_complete; }
+    bool is_active() { return _active; }
+    void deactivate() { _active = false; } // If the copter loses GPS during it's flight, this method should be called to deactivate SafeRTL until disarmed and re-armed.
+    // the two cleanup steps. These should be run regularly, maybe even by a different thread
+    void rdp();
+    void detect_loops();
+private:
+    bool _logging_enabled;
+    bool _active; // if the path becomes too long to keep in memory, and too convoluted to be cleaned up, SafeRTL will be permanently deactivated (for the remainder of the flight)
+    // misc cleanup helper methods:
+    void _reset_rdp();
+    void _reset_pruning();
+    void _zero_points_by_simplification_bitmask();
+    void _remove_unacceptable_overlapping_loops();
+    void _zero_points_by_loops(uint8_t);
+    void _remove_empty_points();
+    // _segment_segment_dist returns two things, the closest distance reached between 2 line segments, and the point exactly between them.
+    typedef struct {
+        float distance;
+        Vector3f point;
+    } dist_point;
+    // typedef struct dist_point dist_point;
+    static dist_point _segment_segment_dist(const Vector3f&, const Vector3f&, const Vector3f&, const Vector3f&);
+    static float _point_line_dist(const Vector3f&, const Vector3f&, const Vector3f&);
+
+    // Simplification state
+    bool _simplification_complete;
+    // structure and buffer to hold the "to-do list" for the RDP algorithm.
+    typedef struct {
+        uint8_t start;
+        uint8_t finish;
+    } start_finish;
+    AP_Buffer<start_finish,RDP_STACK_LEN> _simplification_stack;
+    // the result of the simplification algorithm
+    std::bitset<MAX_PATH_LEN> _simplification_bitmask;
+    // everything before _simplification_clean_until has been calculated already to be un-simplify-able. This avoids recalculating a known result.
+    uint8_t _simplification_clean_until;
+
+    // Pruning state
+    bool _pruning_complete;
+    uint8_t _pruning_current_i;
+    uint8_t _pruning_min_j;
+    typedef struct {
+        uint8_t start_index;
+        uint8_t end_index;
+        Vector3f halfway_point;
+    } loop;
+    // the result of the pruning algorithm
+    std::vector<loop> _prunable_loops; // TODO this might allocate memory while flying. Maybe consider a way to do this with an array or AP_Buffer. or at least use std::list because it is better at handling deletions
```

Comment:
My library allocates memory before arm and does *all* calculations just in place, even without stack usage. And never, never occupies CPU for 100uS :)